### ISSUE TITLE
git-clang-format: InsertBraces on a diff does not apply braces to entire if-else-if-else.

### ISSUE THREAD

**Issue by pbos** [llvm/llvm-project#60153]
From a report in https://bugs.chromium.org/p/chromium/issues/detail?id=1392808#c5 formatting the diff applies braces to an else-if without the previous if or else-if statements:

```
+++ b/third_party/blink/renderer/core/page/focus_controller.cc
@@ -203,9 +203,9 @@ class FocusNavigation : public GarbageCollected<FocusNavigation> {
       owner = node.ParentOrShadowHostElement();
     else if (&node == node.ContainingTreeScope().RootNode())
       owner = TreeOwner(&node);
-    else if (owner_popover_invoker)
+    else if (owner_popover_invoker) {
       owner = owner_popover_invoker;
-    else if (node.parentNode())
+    } else if (node.parentNode())
       owner = FindOwner(*node.parentNode());
```

It would be preferable if we extended diff formatting to the entire if-else-if-else thing. This also unfortunately hits one of our lint scripts:

`third_party/blink/renderer/core/page/focus_controller.cc:206:  If one part of an if-else statement uses curly braces, the other part must too.  [whitespace/braces] [4]`

A known workaround is of course to format the codebase to be consistent with InsertBraces but that's currently controversial on our end.

**Comment by pbos** [comment]
@owenca as this may be something you're interested in?

**Comment by pbos** [comment]
We may also be able to grep for and format most of our existing braceless if/else-ifs, which would minimize the impact of this bug on our end without requiring a complete reformat.

**Comment by owenca** [comment]
This is not a bug IMO. clang-format only formats what's given to it. If git-clang-format tells it to only format a range of lines, it should not touch other lines. Even if it might make sense to insert braces for the entire if-else-if-else chain, clang-format would not be able to only insert braces for the out of range lines. It would have to do other formatting on those lines as well, which would be a real bug.

**Comment by pbos** [comment]
I think that's a reasonable stance, just some maybe unexpected side-effects is that if you run `clang-format` on the diff of the following:

```
if (a)  // Touch only this line.
  b;
else if (c)
  d;
else
  e;
```

Then the first time you format the diff, you touched the `if` so that changes it into `if (a) {` and the `else if` into `} else if`, which touches that line, and then if you format the diff again, that changes `else` into `} else`, and finally you format that.

Because our tools assume that what you upload should be the same as if you've formatted it, it'll run `clang-format` on the diff and complain if there are any changes, so we end up in a space where you run `git cl format` over and over again until it's formatted all the way to the else. Just thought I'd share that. It's not unreasonable to keep `clang-format` unaware of how the diff will be used. If we wanted to fix it maybe we'd just retry formatting until it's stable or give up after N invocations.

**Comment by owenca** [comment]
I see. Unfortunately, there's no simple solution as far as I can tell.

**Comment by pkasting** [comment]
FWIW, I am continually bitten by cases where I touch the first line here:
```
  if (cond)
    xyz;
  abc;
```
...and have it change to:
```
  if (cond) {
    xyz;
  abc;
```

Note the lack of closing brace. This of course causes compile errors... most of the time. I am scared I'll introduce a subtle bug.

This problem bites me something like once a week and has done so as recently as last week.

**Comment by owenca** [comment]
Which version of clang-format? Please post a reproducer.

**Comment by pkasting** [comment]
```
$ clang-format --version
clang-format version 18.0.0 (https://chromium.googlesource.com/a/external/github.com/llvm/llvm-project 10664813fca8d5ccbfd90bae9e791b7062dabd7c)
```

My repro steps were:
* In a Chromium checkout on Windows, open base/time/time.cc in Visual Studio 2022 17.8.6. Ensure Visual Studio is configured to use clang-format with style "Chromium" and a custom executable location of $YOUR_CHROMIUM_SRC_DIR/buildtools/win/clang-format.exe, and to automatically format on paste.
* In the first conditional in TimeDelta::CeilToMultiple(), paste "!a || " into the middle of the conditional.
* Note that the end of the line gains a '{', but no closing brace is added after the conditional body.

Here's what that code looks like: https://source.chromium.org/chromium/chromium/src/+/main:base/time/time.cc;l=52;drc=e5fff99cbb0380ea7f44c60ee15554b6b56320fb
```
TimeDelta TimeDelta::CeilToMultiple(TimeDelta interval) const {
  if (is_inf() || interval.is_zero())
    return *this;
  const TimeDelta remainder = *this % interval;
  ...
```

If instead I open the file in some editor that has no clang-format integration, touch that line, then run `git cl format` (which uses Chromium's git-cl tool to invoke clang-format), I get a closing brace. So, obviously, `git cl format` and devenv.exe are invoking clang-format differently here. But normally devenv.exe's invocation of clang-format is willing to reformat code on nearby lines when you paste into a line, so I'm not sure precisely what the difference is.

**Comment by owenca** [comment]
Can you also post your .clang-format file?

**Comment by pkasting** [comment]
Chromium's .clang-format is at https://chromium.googlesource.com/chromium/src/+/main/.clang-format .

**Comment by owenca** [comment]
I can't reproduce it by running clang-format directly:
```
$ diff time.cc foo.cc
52c52
<   if (is_inf() || interval.is_zero())
---
>   if (is_inf() || !a || interval.is_zero())
$ clang-format -version
clang-format version 18.1.8
$ clang-format -lines=52:52 foo.cc | diff foo.cc -
52c52
<   if (is_inf() || !a || interval.is_zero())
---
>   if (is_inf() || !a || interval.is_zero()) {
53a54
>   }
$ 
```

**Comment by pkasting** [comment]
Yes, I think the Visual Studio invocation of clang-format is critical here, but I don't know exactly how Visual Studio is invoking, or how to figure out. Maybe some kind of judicious use of ProcessExplorer would sniff this out...
